File: Intersection of Two Arrays II .java
class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {

        if(nums2.length>nums1.length) intersect(nums2,nums1);

        HashMap<Integer,Integer> map = new HashMap<>();

        ArrayList<Integer> res = new ArrayList<>();

       for(int i=0 ; i<nums1.length ; i++)
       {
           map.put(nums1[i] , map.getOrDefault(nums1[i],0)+1);
       }

        for(int n :nums2)
        {
            if(map.containsKey(n))
            {
                res.add(n);
                map.put(n , map.getOrDefault(n,0)-1);
                if(map.get(n)==0) map.remove(n);
            }
        }

        int[] ans = new int[res.size()];
        int index=0;

        for(int n : res)
        {
            ans[index++] = n;
        }

        return ans;

    }
}

File: Intersection_of_Two_ArraysII.py
# Time COmplexity : O(n + m), where n and m are the lengths of nums1 and nums2, respectively
# Space COmplexity : O(n + m)
from collections import Counter
from typing import List

class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        counter1 = Counter(nums1)
        counter2 = Counter(nums2)
        
        result = []
        for num, count in counter1.items():
            if num in counter2:
                result.extend([num] * min(count, counter2[num]))
        
        return result